src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -108,38 +106,48 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     		}
     	} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
-			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
-				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
-				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+			Locale locale = Locale.forLanguageTag(dto.getLocale());
+			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
+				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
+					cacheItem = getMessagesFallbackLocale(fallbackLocalesIter);
+			} else { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(ps.getSupportedLocales(), locale);
+				if (matchedLocale != null) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
 					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
-					return new ComponentService(matchedLocaleDTO).getMessages();
-				}
+					cacheItem = new ComponentService(matchedLocaleDTO).getMessages();
+				} else  // Requested locale is not supported and does not match any supported locales
+					cacheItem = getMessagesFallbackLocale(fallbackLocalesIter);
```

Comment:
I think if you handle locale before looking for the messages from cache, the logic here may be  simplifier. 

In my opinion, the logic should be as below:
1. Handle locale:
Locale matchedLocale = LocaleUtility.pickupLocaleFromList(ps.getSupportedLocales(), locale);
2. If matchedLocale is not empty, use matchedLocale to fetch messages from data source;
if matchedLocale is empty, directly use fallback locale to fetch messages from data source;
3. if above fetch failed, fallback;

Reply:
I understand what you are saying. The problem with this is due to the locale being part of the DTO. If the matched locale is not null, there are 2 possible scenarios:

a. The requested locale is a supported locale (exact match) -- lines 110-113
b. The requested locale is not an exact match (e.g. fr-CA which is not supported matches fr which is supported) -- lines 117-118

If it is (a), createCacheItem in line 111 fetches the matched locale and caches it. Take note that it uses the **dto's locale in the cache key**.
If it is (b), createCacheItem must be done on the supported locale (fr), not the requested locale (fr-CA). If I directly call createCacheItem for this case, the cache key will have fr-CA for the same set of fr messages. causing duplicate sets of messages in the cache, which is not good.

Reply:
Can reseting dto's locale after locale matching resolve the problem you mentioned?

I think when matchedLocale is not empty, all later logic(no matter find from cache or find from data source) should use matchedLocale, request locale is useless.